snowflake-api/src/lib.rs

```diff
@@ -225,22 +245,53 @@ impl SnowflakeApi {
             .with_token(info.creds.aws_token)
             .build()?;
 
+        let s3_mutex = Arc::new(Mutex::new(s3));
+
         // todo: security vulnerability, external system tells you which local files to upload
+        let mut final_source_locs: Vec<String> = Vec::new();
         for src_path in src_locations {
-            let path = Path::new(src_path);
-            let filename = path
-                .file_name()
-                .ok_or(SnowflakeApiError::InvalidLocalPath(src_path.clone()))?;
-
-            // fixme: unwrap
-            let dest_path = format!("{}{}", bucket_path, filename.to_str().unwrap());
-            let dest_path = object_store::path::Path::parse(dest_path)?;
-
-            let src_path = object_store::path::Path::parse(src_path)?;
+            for entry in glob::glob(src_path).expect("Failed to read glob pattern") {
+                match entry {
+                    Ok(path) => match path.to_str() {
+                        Some(item) => final_source_locs.push(item.to_string()),
+                        None => return Err(SnowflakeApiError::InvalidLocalPath(src_path.clone())),
+                    },
+                    Err(_) => return Err(SnowflakeApiError::InvalidLocalPath(src_path.clone())),
+                }
+            }
+        }
 
-            let fs = LocalFileSystem::new().get(&src_path).await?;
+        let bucket_path = bucket_path.to_string();
+        let mut set = JoinSet::new();
+        for src_path in final_source_locs {
+            let mutex1 = Arc::clone(&s3_mutex);
+            let bucket_path = bucket_path.clone();
+            while set.len() >= 5 {
+                set.join_next().await.unwrap().unwrap();
```

Comment:
So only the last batch of 5 will get proper error handling and this crashes with exception? I think there could be nicer way to split tasks into batches of given size.

Reply:
This doesnt seem to offer a performance improvement. I will just remove it for now.